File: core-platform/apps/core-webapp/src/services/bug.service.ts
import axiosInstance from "../axiosInstance";
import type {
  BugDTO,
  BugPageResponse,
  CreateBugRequest,
  UpdateBugRequest,
} from "../types/bug.types";

const BUG_API_BASE = "/bugs";

export const bugService = {
  // CRUD Operations
  createBug: async (data: CreateBugRequest): Promise<BugDTO> => {
    const response = await axiosInstance.post(BUG_API_BASE, data);
    return response.data.data;
  },

  updateBug: async (id: number, data: UpdateBugRequest): Promise<BugDTO> => {
    const response = await axiosInstance.put(`${BUG_API_BASE}/${id}`, data);
    return response.data.data;
  },

  deleteBug: async (id: number): Promise<void> => {
    await axiosInstance.delete(`${BUG_API_BASE}/${id}`);
  },

  getBugById: async (
    id: number,
    includeDetails: boolean = false,
  ): Promise<BugDTO> => {
    const response = await axiosInstance.get(`${BUG_API_BASE}/${id}`, {
      params: { includeDetails },
    });
    return response.data.data;
  },

  // Search with pagination
  searchBugs: async (
    organizationId: number,
    keyword?: string,
    page: number = 0,
    size: number = 20,
  ): Promise<BugPageResponse> => {
    const response = await axiosInstance.get(`${BUG_API_BASE}/search`, {
      params: {
        organizationId,
        keyword,
        page,
        size,
        sort: "createdAt,desc",
      },
    });
    return response.data.data;
  },

  // Filters
  getBugsByProject: async (projectId: number): Promise<BugDTO[]> => {
    const response = await axiosInstance.get(
      `${BUG_API_BASE}/project/${projectId}`,
    );
    return response.data.data;
  },

  getBugsByAssignee: async (userId: number): Promise<BugDTO[]> => {
    const response = await axiosInstance.get(
      `${BUG_API_BASE}/assignee/${userId}`,
    );
    return response.data.data;
  },

  // Status Management
  changeStatus: async (id: number, newStatus: string): Promise<BugDTO> => {
    const response = await axiosInstance.put(
      `${BUG_API_BASE}/${id}/status`,
      null,
      {
        params: { newStatus },
      },
    );
    return response.data.data;
  },

  changeSeverity: async (id: number, newSeverity: string): Promise<BugDTO> => {
    const response = await axiosInstance.put(
      `${BUG_API_BASE}/${id}/severity`,
      null,
      {
        params: { newSeverity },
      },
    );
    return response.data.data;
  },

  // Task Linking
  linkToTask: async (id: number, taskId: number): Promise<void> => {
    await axiosInstance.put(`${BUG_API_BASE}/${id}/link-task/${taskId}`);
  },

  unlinkFromTask: async (id: number): Promise<void> => {
    await axiosInstance.put(`${BUG_API_BASE}/${id}/unlink-task`);
  },

  // Close/Reopen
  closeBug: async (id: number): Promise<void> => {
    await axiosInstance.put(`${BUG_API_BASE}/${id}/close`);
  },

  reopenBug: async (id: number): Promise<void> => {
    await axiosInstance.put(`${BUG_API_BASE}/${id}/reopen`);
  },
};

export default bugService;

File: core-platform/services/core-service/src/main/java/com/dev/core/service/bug/BugService.java
package com.dev.core.service.bug;

import com.dev.core.model.bug.BugDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface BugService {

    BugDTO createBug(BugDTO dto);

    BugDTO updateBug(Long id, BugDTO dto);

    void deleteBug(Long id);

    BugDTO getBugById(Long id, boolean includeDetails);

    Page<BugDTO> searchBugs(Long organizationId, String keyword, Pageable pageable);

    List<BugDTO> getBugsByProject(Long projectId);

    List<BugDTO> getBugsByAssignee(Long userId);
    
    List<BugDTO> getMyBugs(Long organizationId);

    BugDTO changeBugStatus(Long bugId, String newStatus);

    BugDTO changeBugSeverity(Long bugId, String newSeverity);

    void reopenBug(Long bugId);

    void verifyBug(Long bugId, Long verifiedBy);

    void closeBug(Long bugId);

    void escalateBug(Long bugId, String reason);

    void linkBugToTask(Long bugId, Long taskId);

    void unlinkBugFromTask(Long bugId);
}
